build_datacard.py

```diff
@@ -173,99 +181,101 @@ def apply_selection(cols,year):
         return outfiles1+outfiles2
 
     common.logger.info(f'Selection: {selection}')
-    central = apply_selection(central,year)
-    mt = central.to_numpy(['mt']).ravel()
-
-    if metadata["sample_type"]=="data":
-        w = None
-    else:
-        w = central.to_numpy(['puweight']).ravel()
-        event_weight = common.get_event_weight(central,lumi)
+    cen_columns = apply_selection(cen_columns,year)
+    cen_array = cen_columns.to_numpy([hist_var]).flatten()
+    w = None
+    if metadata["sample_type"] != "data":
+        w = cen_columns.to_numpy(['puweight']).ravel()
+        event_weight = common.get_event_weight(cen_columns,lumi)
         w *= event_weight
         # save event weight
         metadata['event_weight'] = common.get_single_event_weight(event_weight)
-    mth_central = common.MTHistogram(mt, w)
-    mth_central.metadata.update(metadata)
-    mths['central'] = mth_central
+
+    # Defining the histogram type to use
+    VarHistogram = common.registered_varhists[hist_var]
+
+    hist_central = VarHistogram(cen_array, w)
+    hist_central.metadata.update(metadata)
+    hist_variants['central'] = hist_central
 
     if metadata["sample_type"]=="sig":
         # Scale
         good_scales = np.array([0,1,2,3,4,6,8])
-        scale_weight = central.to_numpy(['scaleweights'])
+        scale_weight = cen_columns.to_numpy(['scaleweights'])
         scale_weight = scale_weight[ak.num(scale_weight,axis=1)>np.max(good_scales)]
         scale_weight = scale_weight[:, good_scales]
         if len(scale_weight):
-            weight_up = w * np.max(scale_weight, axis=-1) * central.metadata['scale_factor_up']
-            weight_down = w * np.min(scale_weight, axis=-1) * central.metadata['scale_factor_down']
-            mths['scale_up'] = common.MTHistogram(mt, weight_up)
-            mths['scale_down'] = common.MTHistogram(mt, weight_down)
+            weight_up = w * np.max(scale_weight, axis=-1) * cen_columns.metadata['scale_factor_up']
+            weight_down = w * np.min(scale_weight, axis=-1) * cen_columns.metadata['scale_factor_down']
+            hist_variants['scale_up'] = VarHistogram(cen_array, weight_up)
+            hist_variants['scale_down'] = VarHistogram(cen_array, weight_down)
 
         # JEC/JER/JES
         def mth_jerjecjes(tag):
             col = svj.Columns.load(get_variation(tag))
             col = apply_selection(col,year)
-            mt = col.to_numpy(['mt']).flatten()
+            var_arr = col.to_numpy([hist_var]).flatten()
```

Comment:
the potential limitation here is if we want to fill a histogram that is a computation based on multiple arrays, rather than just a single array. i was envisioning the histogram class having a "fill" function that took the entire `col` as input, and then each derived histogram class would know how to extract its output (which could be a default operation like this in the base class, and then only overridden in special cases).

Reply:
One impact this might have is that the same `col.to_numpy` might be called multiple times. I'm not sure if this would have a significant performance impact?

Reply:
I don't think it's worth worrying about that unless we see that it causes a noticeable problem in practice. By default, I think it is still better to convert on demand rather than converting everything in advance (which has a significant memory implication).